docs/plugins/development/creating-text-analysis-plugins.asciidoc

```diff
@@ -0,0 +1,129 @@
+of component names and implementation classes.
+
+Note that only JAR files at the root of the plugin are added to the classpath
+for the plugin. If you need other resources, package them into a resources JAR.
+
+[discrete]
+==== The analysis plugin API
+
+Text analysis plugins can be written against the analysis plugin API. This API
+includes:
+
+* the entirety of the core Lucene module
+* the {es} `plugin-api`, which includes important annotations for the plugin 
+framework
+* `plugin-analysis-api`, which contains the classes that an analysis plugin can 
+implement.
+
+The advantage of using this API is that the resulting plugin is
+forward-compatible with new minor versions of {es}. While elements may be added
+to the API, existing element will not be changed or removed, except in a new
+major release. In other words, once you have a working artifact, you can re-use
+it when you upgrade {es} to a new bugfix or minor version.
+
+It's also possible to create text analysis plugins using the method described in
+<<creating-non-text-analysis-plugins>>. Howeverm plugins created using that 
+method are pinned to specific version of {es}. You will need to recompile them
+when upgrading{es}.
+
+A text analysis can implement four factory classes that are provided by the 
+analysis plugin API.
+
+* `AnalyzerFactory` to create a Lucene analyzer
+* `CharFilterFactory` to create a character character filter
+* `TokenFilterFactory` to create a Lucene token filter
+* `TokenizerFactory` to create a Lucene tokenizer
+
+The key to implementing a text analysis plugin is the `@NamedComponent`
+annotation. Many of {es}'s components have names that are used in
+configurations. For example, the keyword analyzer is referenced in configuration
+with the name `"keyword"`. Once your custom plugin is installed in your cluster,
+your named components may be referenced by name in these configurations as well.
+
+[discrete]
+==== Development process
+
+Elastic provides a Grade plugin, `elasticsearch.stable-esplugin`, that makes it
+easier to develop and package plugins. The steps below assume you use this 
+plugin. However, you don't need Gradle to create plugins.
+
+The {es} Github repository contains
+{es-repo}tree/main/plugins/examples/stable-analysis[an example analysis plugin].
+The example `build.gradle` build script provides a good starting point for 
+developing your own plugin.
+
+[discrete]
+===== Prerequisites
+
+Plugins are written in Java, so you need to install a Java Development Kit
+(JDK). Install Gradle if you want to use Gradle.
+
+[discrete]
+===== Step by step
+
+. Create a directory for your project. 
+. Copy the example `build.gradle` build script to your project directory.  Note
+that this build script uses the `elasticsearch.stable-esplugin` gradle plugin to
+build your plugin.
+. Edit the `build.gradle` build script:
+** Add a definition for the `elasticsearchVersion` and matching `luceneVersion` 
```

Comment:
we use elasticsearchVersion in our examples too. but this made me realise that maybe we should rename this to pluginApiVersion?

Reply:
I think that makes sense. The instructions now recommend that folks copy `build.gradle` from the example, so we'd have to update the example.

Reply:
let me create a PR to rename this in plugins/examples

Reply:
https://github.com/elastic/elasticsearch/pull/93501